From 30ad6c7fcc2edb0a798e2ec0da5ea6ec68a19505 Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@notvalid.com>
Date: Wed, 14 Oct 2026 08:42:58 -0700
Subject: [PATCH] Adapt pinned Guix build to headless Lotus with independent
 Chronik and NNG dependencies

Use the validated Rust 1.93.0 toolchain and current Lotus CMake target and option names. Build and run C++ checks before security and symbol checks, package daemon and CLI with upstream COPYING, and avoid GUI, wallet, ZeroMQ and proof-manager targets. Pass the depends toolchain explicitly to the existing NNG and Flatbuffers recipes. Initial validation is limited to x86_64-linux-gnu; no macOS ARM cross-build claim is made.
---
 contrib/guix/guix-build           |  8 ++++--
 contrib/guix/libexec/build.sh     | 47 ++++++++++++++++++++-----------
 contrib/guix/libexec/prelude.bash |  2 +-
 depends/Makefile                  |  2 ++
 depends/packages/nng.mk           |  4 +++
 5 files changed, 43 insertions(+), 20 deletions(-)

diff --git a/contrib/guix/guix-build b/contrib/guix/guix-build
index 3b141f6109..ccac864c61 100755
--- a/contrib/guix/guix-build
+++ b/contrib/guix/guix-build
@@ -91,7 +91,7 @@ fi
 
 # Default to building for all supported HOSTs (overridable by environment)
 # TODO: support arm64-apple-darwin,
-export HOSTS="${HOSTS:-x86_64-linux-gnu x86_64-w64-mingw32 x86_64-apple-darwin aarch64-linux-gnu}"
+export HOSTS="${HOSTS:-x86_64-linux-gnu}"
 
 # Usage: distsrc_for_host HOST
 #
@@ -257,7 +257,7 @@ mkdir -p "$OUTDIR_BASE"
 # Download the depends sources now as we won't have internet access in the build
 # container
 for host in $HOSTS; do
-    make -C "${PWD}/depends" -j"$JOBS" download-"$(host_to_commonname "$host")" ${V:+V=1} ${SOURCES_PATH:+SOURCES_PATH="$SOURCES_PATH"}
+    make -C "${PWD}/depends" -j"$JOBS" HOST="$host" download-one NO_QT=1 NO_QR=1 NO_ZMQ=1 NO_WALLET=1 BUILD_CPYTHON=1 ${V:+V=1} ${SOURCES_PATH:+SOURCES_PATH="$SOURCES_PATH"}
 done
 
 # Usage: outdir_for_host HOST SUFFIX
@@ -415,6 +415,8 @@ EOF
                                  -- env HOST="$host" \
                                         DISTNAME="$DISTNAME" \
                                         JOBS="$JOBS" \
+                                        LOTUS_CHRONIK="${LOTUS_CHRONIK:-ON}" \
+                                        LOTUS_NNG="${LOTUS_NNG:-ON}" \
                                         SOURCE_DATE_EPOCH="${SOURCE_DATE_EPOCH:?unable to determine value}" \
                                         ${V:+V=1} \
                                         ${SOURCES_PATH:+SOURCES_PATH="$SOURCES_PATH"} \
@@ -428,7 +430,7 @@ EOF
 
 done
 
-cat ${OUTDIR_BASE}/**/SHA256SUMS.part > ${OUTDIR_BASE}/SHA256SUMS
+(cd "$OUTDIR_BASE" && find . -type f -name "*.tar.gz" -print0 | sort -z | xargs -0 sha256sum > SHA256SUMS)
 echo "---8<----------------------8<------------------------8<---"
 cat ${OUTDIR_BASE}/SHA256SUMS
 echo "---8<----------------------8<------------------------8<---"
diff --git a/contrib/guix/libexec/build.sh b/contrib/guix/libexec/build.sh
index f718883c80..1217fafee6 100755
--- a/contrib/guix/libexec/build.sh
+++ b/contrib/guix/libexec/build.sh
@@ -46,7 +46,14 @@ OUTDIR="${DISTSRC}/output"
 
 # The depends folder also serves as a base-prefix for depends packages for
 # $HOSTs after successfully building.
-BASEPREFIX="${PWD}/depends"
+export BASEPREFIX="${PWD}/depends"
+# This maintained candidate initially validates only native Linux x86_64.
+[[ "$HOST" == x86_64-linux-gnu ]] || { echo "Unsupported Lotus Guix host: $HOST" >&2; exit 1; }
+export CMAKE_TOOLCHAIN_FILE="${PWD}/cmake/platforms/Linux64.cmake"
+LOTUS_CHRONIK="${LOTUS_CHRONIK:-ON}"
+LOTUS_NNG="${LOTUS_NNG:-ON}"
+NO_NNG=
+[[ "$LOTUS_NNG" == ON ]] || NO_NNG=1
 
 # Given a package name and an output name, return the path of that output in our
 # current guix environment
@@ -156,13 +163,7 @@ export GUIX_LD_WRAPPER_DISABLE_RPATH=yes
 # Determine the correct value for -Wl,--dynamic-linker for the current $HOST
 case "$HOST" in
     *linux*)
-        glibc_dynamic_linker=$(
-            case "$HOST" in
-                x86_64-linux-gnu)      echo /lib64/ld-linux-x86-64.so.2 ;;
-                aarch64-linux-gnu)     echo /lib/ld-linux-aarch64.so.1 ;;
-                *)                     exit 1 ;;
-            esac
-        )
+        glibc_dynamic_linker=/lib64/ld-linux-x86-64.so.2
         ;;
 esac
 
@@ -175,7 +176,8 @@ export TZ="UTC"
 ####################
 
 # Build the depends tree, overriding variables that assume multilib gcc
-make -C depends --jobs="$JOBS" HOST="$HOST" \
+make -C depends --jobs="$JOBS" HOST="$HOST" NO_QT=1 NO_QR=1 NO_ZMQ=1 NO_WALLET=1 BUILD_CPYTHON=1 NO_NNG="$NO_NNG" NO_FLATBUFFERS="$NO_NNG" \
+                                   CMAKE_TOOLCHAIN_FILE="$CMAKE_TOOLCHAIN_FILE" BASEPREFIX="$BASEPREFIX" \
                                    ${V:+V=1} \
                                    ${SOURCES_PATH+SOURCES_PATH="$SOURCES_PATH"} \
                                    ${BASE_CACHE+BASE_CACHE="$BASE_CACHE"} \
@@ -215,7 +217,7 @@ esac
 curl -sSf https://static.rust-lang.org/rustup/archive/1.28.2/x86_64-unknown-linux-gnu/rustup-init -o rustup-init
 echo "20a06e644b0d9bd2fbdbfd52d42540bdde820ea7df86e92e533c073da0cdd43c rustup-init" | sha256sum -c
 chmod +x rustup-init
-./rustup-init -y --default-toolchain=1.87.0
+./rustup-init -y --default-toolchain=1.93.0
 rm ./rustup-init
 # shellcheck disable=SC1091
 source "$HOME/.cargo/env"
@@ -232,6 +234,7 @@ cmake -GNinja .. \
     -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TOOLCHAIN_FILE} \
     -DBUILD_WALLET=OFF \
     -DBUILD_CHRONIK=OFF \
+    -DBUILD_BITCOIN_NNG="$LOTUS_NNG" \
     -DBUILD_QT=OFF \
     -DBUILD_ZMQ=OFF \
     -DENABLE_QRCODE=OFF \
@@ -362,8 +365,18 @@ mkdir -p "$DISTSRC"
       -DENABLE_REDUCE_EXPORTS=ON \
       -DCMAKE_INSTALL_PREFIX="${INSTALLPATH}" \
       -DCCACHE=OFF \
-      -DBUILD_CHRONIK=ON \
-      -DBUILD_PROOF_MANAGER_CLI=ON \
+      -DBUILD_CHRONIK="$LOTUS_CHRONIK" \
+      -DBUILD_PROOF_MANAGER_CLI=OFF \
+      -DRust_TOOLCHAIN=1.93.0 \
+      -DBUILD_QT=OFF \
+      -DBUILD_ZMQ=OFF \
+      -DBUILD_BITCOIN_NNG="$LOTUS_NNG" \
+      -DBUILD_WALLET=OFF \
+      -DBUILD_SEEDER=OFF \
+      -DBUILD_IGUANA=OFF \
+      -DBUILD_TX=OFF \
+      -DENABLE_UPNP=OFF \
+      -DENABLE_NATPMP=OFF \
       "${CMAKE_EXTRA_OPTIONS[@]}" \
       "${CMAKE_C_FLAGS}" \
       "${CMAKE_CXX_FLAGS}" \
@@ -382,7 +395,10 @@ EOF
     fi
 
     # Build Bitcoin ABC
-    ninja
+    ninja -j"$JOBS" lotusd lotus-cli test_lotus
+    ninja -j"$JOBS" check-bitcoin
+    ./src/lotusd --version
+    ./src/lotus-cli --version
     ninja security-check
     ninja symbol-check
 
@@ -402,8 +418,8 @@ EOF
             popd
             ;;
         *linux*)
-            ninja install-debug
-            ninja install-proof-manager-cli
+            ninja install-lotus-binaries-debug
+            install -m 644 ../COPYING "${INSTALLPATH}/COPYING"
             pushd installed
             find ${DISTNAME} -not -name "*.dbg" -print0 | sort --zero-terminated | tar --create --no-recursion --mode='u+rw,go+r-w,a+X' --null --files-from=- | gzip -9n > ${OUTDIR}/${DISTNAME}-${HOST}.tar.gz
             find ${DISTNAME} -name "*.dbg" -print0 | sort --zero-terminated | tar --create --no-recursion --mode='u+rw,go+r-w,a+X' --null --files-from=- | gzip -9n > ${OUTDIR}/${DISTNAME}-${HOST}-debug.tar.gz
@@ -426,7 +442,6 @@ EOF
             ninja osx-zip
             mv "${OSX_VOLNAME}.zip" ${OUTDIR}/${DISTNAME}-osx-unsigned.zip
 
-            ninja install-proof-manager-cli
 
             pushd installed
             find . -path "*.app*" -type f -executable -exec mv {} ${DISTNAME}/bin/bitcoin-qt \;
diff --git a/contrib/guix/libexec/prelude.bash b/contrib/guix/libexec/prelude.bash
index e6cb64b439..f0b021157c 100644
--- a/contrib/guix/libexec/prelude.bash
+++ b/contrib/guix/libexec/prelude.bash
@@ -50,7 +50,7 @@ fi
 # across time.
 time-machine() {
     # shellcheck disable=SC2086
-    guix time-machine --url=https://git.savannah.gnu.org/git/guix.git \
+    guix time-machine --url=https://codeberg.org/guix/guix.git \
                       --commit=c5eee3336cc1d10a3cc1c97fde2809c3451624d3 \
                       --cores="$JOBS" \
                       --keep-failed \
diff --git a/depends/Makefile b/depends/Makefile
index 96b06dadea..d77de7d7ac 100644
--- a/depends/Makefile
+++ b/depends/Makefile
@@ -182,6 +182,8 @@ endif
 ifeq ($(cpython_packages_),)
 native_packages += $(cpython_native_packages)
 packages += $(cpython_packages)
+endif
+
 ifneq ($(nng_packages_),)
 packages += $(nng_packages)
 endif
diff --git a/depends/packages/nng.mk b/depends/packages/nng.mk
index b723810683..46e9771817 100644
--- a/depends/packages/nng.mk
+++ b/depends/packages/nng.mk
@@ -15,6 +15,10 @@ define $(package)_config_cmds
     -DCMAKE_BUILD_TYPE=Release \
     -DCMAKE_CROSSCOMPILING=on \
     -DNNG_ENABLE_NNGCAT=off \
+    -DNNG_TESTS=OFF \
+    -DNNG_TOOLS=OFF \
+    -DBUILD_SHARED_LIBS=OFF \
+    -DCMAKE_POSITION_INDEPENDENT_CODE=ON \
     -DCMAKE_TOOLCHAIN_FILE=$(CMAKE_TOOLCHAIN_FILE) \
     -DBASEPREFIX=$(BASEPREFIX)
 endef
-- 
2.50.1 (Apple Git-155)

